### PR TITLE
Initial rewrite upload

### PR DESCRIPTION
**Pull request by oliverfoster**
Extended considerably.

Allows for trickle on article and block.
Huge variation in settings.
Tutor compatible.
Tested.

Original repo: https://github.com/cgkineo-projects/adapt-contrib-trickle/tree/trickle-rewrite
Example: http://rawgit.com/oliverfoster/builds/trickle-rewrite/index.html

2015-05-21: Updated example to match current develop course - included PRs:

https://github.com/adaptlearning/adapt-contrib-assessmentResults/pull/4
https://github.com/adaptlearning/adapt-contrib-assessmentResultsTotal/pull/3

### CODE REVIEW

**Comment by dancgray** [comment]
Cool. Thanks for submitting this Ollie, we use trickle quite a lot and have numerous issues with it.

Will test this extensively.

**Comment by oliverfoster** [comment]
Awesome, be prepared for the results components to be hidden on a page revisit at the moment. I've got to sort that out.

**Comment by oliverfoster** [comment]
Just changed the example to have one normal looking trickle button. Give it some minutes to update.

**Comment by moloko** [comment]
Code's looking really good, much clearer than any previous iteration I've seen

There are still a few functions that could do with an explanatory comment - which is unavoidable in something of this complexity. I'll try and look through over the next couple of days and state where I think a bit more explanation could help.

I also think it would benefit from some high-level technical documentation and perhaps also some more detailed documentation for users on the configuration settings. All that can come once it's actually in though.

**Comment by oliverfoster** [comment]
Cool. Good plan.

**Comment by zenduo** [comment]
This looks really good :) will give it a go.

**Comment by zenduo** [comment]
Very nice update...

2 things that i spotted while testing:

_duration didn't work for me, rename to scrollDuration as you suggested
_onChildren when used on an article, the default value if not set, should be true in my opinion

**Comment by dancgray** [comment]
Nice, seems to fix an issue with trickle on return to narrative when the stage is > 0.

Getting an issue is you set trickle on the last article of a page. Continue button is show after completing the last component at the end of the page. Should not be shown if there is no other content on the page.

Error:
Uncaught TypeError: Cannot read property 'get' of undefined adapt-contrib-trickle.js:515
JSON used on article "_id":"a-10":
"_trickle": {
    "_isEnabled": true,
    "_onChildren": true 
}

**Comment by oliverfoster** [comment]
Cool, will fix

**Comment by oliverfoster** [comment]
That should do it guys. If you could retest that'd be awesome.

File: js/Utility/Models.js [hunk 1]
```diff
@@ -0,0 +1,132 @@
...
+*/
+
+define([
+    'coreJS/adapt',
+    '../DataTypes/StructureType'
+], function(Adapt, StructureType) {
+
+    var ModelUtilities = {
+
+        getDescendantsFlattened: function(id, parentFirst) {
```
Comment (moloko) [inline js/Utility/Models.js]: could use a brief comment explaining bit more about what this does and why/when it is needed

File: js/Utility/Models.js [hunk 2]
```diff
@@ -0,0 +1,132 @@
...
+                    if (parentFirst != true) descendants.push(child);
+
+                }
+
+            }
+
+            return new Backbone.Collection(descendants);
+        },
+
+        findRelative: function(model, relativeString) {
```
Comment (moloko) [inline js/Utility/Models.js]: small explanation of how the 'relative' part of this works would help

File: js/adapt-contrib-trickle.js [hunk 3]
```diff
@@ -1,258 +1,558 @@
...
+        },
+
+        checkResetChildren: function() {
+            var descendantsChildrenFirst = this._descendantsChildrenFirst;
+            for (var i = 0, model; model = descendantsChildrenFirst.models[i++];) {
+                this.checkResetModel(model);
+            }
+        },
+
+        checkResetModel: function(model) {
```
Comment (moloko) [inline js/adapt-contrib-trickle.js]: a brief overview of the logic here would help

File: js/adapt-contrib-trickle.js [hunk 4]
```diff
@@ -1,258 +1,558 @@
...
+
+            var scrollTo = trickleConfig._scrollTo;
+            this.relativeScrollTo( model, scrollTo );
+        },
+
+        isStepUnlockWaiting: function() {
+            return this._waitForUnlockRequestsCount > 0;
+        },
+        
+        relativeScrollTo: function(model, scrollTo) {
```
Comment (moloko) [inline js/adapt-contrib-trickle.js]: a brief explanation of the 'relative' aspect of this would help

File: js/adapt-contrib-trickle.js (hunk 3 above)
Comment (oliverfoster) [inline js/adapt-contrib-trickle.js]: for the weird for loop?

**Comment by dancgray** [comment]
Update fixes last article/block on page issue.

Setting _stepLocking. _isEnabled = false seems to be the same as _trickle._isEnabled = false. I was expecting it to allow autoscroll without step locking but this does not seem to be the case.

**Comment by oliverfoster** [comment]
There is a small caveat to that one Dan. The _isFullWidth: true setting forces _stepLocking._isEnabled: true - otherwise we end up with lots of fixed-bottom buttons stacked on top of each other. You'll need to turn off _isFullWidth to see it without steplocking.

**Comment by dancgray** [comment]
Thanks Ollie, that seems to work.

+1 from me

File: js/trickle-buttonView.js [hunk 5]
```diff
@@ -0,0 +1,302 @@
...
+                    duration: 250,
+                    complete: _.bind(function() {
+                        if (!bool) this.$('.component-inner').css("visibility", "hidden");
+                    }, this)
+                })
+            }
+            
+        },
+
+        isOnCompleteVisible: function() {
```
Comment (moloko) [inline js/trickle-buttonView.js]: could use a little explanation

File: js/adapt-contrib-trickle.js (hunk 3 above)
Comment (moloko) [inline js/adapt-contrib-trickle.js]: no, of the checkResetModel function. sorry, Github gives you a really odd preview there, the comment was actually related to L147

**Comment by oliverfoster** [comment]
Could we retest and +1 if acceptable?

**Comment by dancgray** [comment]
+1

**Comment by dancgray** [comment]
+1

**Comment by moloko** [comment]
+1

**Comment by dancgray** [comment]
+1

**Comment by taylortom** [comment]
+1

**Pull request merged**